ocaml/xenopsd/lib/xenops_server.ml

```diff
@@ -2726,7 +2726,35 @@ and perform_exn ?subtask ?result (op : operation) (t : Xenops_task.task_handle)
             ) ;
             debug "VM.migrate: Synchronisation point 1"
           in
+          let clean_src_vm () =
+            debug
+              "VM.migrate: clean up src vm before allowing destination to \
+               proceed" ;
+            (* cleanup tmp src VM *)
+            let atomics =
+              [
+                VM_hook_script_stable
+                  ( id
+                  , Xenops_hooks.VM_pre_destroy
+                  , Xenops_hooks.reason__suspend
+                  , new_src_id
+                  )
+              ]
+              @ atomics_of_operation (VM_shutdown (new_src_id, None))
+              @ [
+                  VM_hook_script_stable
+                    ( id
+                    , Xenops_hooks.VM_post_destroy
+                    , Xenops_hooks.reason__suspend
+                    , new_src_id
+                    )
+                ; VM_remove new_src_id
+                ]
+            in
+            perform_atomics atomics t
+          in
           let final_handshake () =
+            clean_src_vm () ;
```

Comment:
As this is not part of the handshake, I prefer the make this call separately in the sequence below (L2797 and L2814).
I'm actually surprised that this hasn't caused problems for plain, non-storage live migration on SMAPIv3 SRs... or has it?

Reply:
Well this will only cause problems if there is an active mirror datapath, which will give you a VDI in use if the destination VM attempts to activate on that active datapath currently being used by mirroring, as designed. If there is no storage migration, then there is no active mirror datapath, and should be no problem.